### DISCUSSION TITLE
Locally safelist classes in component

### DISCUSSION THREAD

**Discussion by benjaminpreiss** [tailwindlabs/tailwindcss#12209] [Help] [answered]
I wonder whether it is possible to safelist classes locally in a (svelte) component.

So that only when that component is imported somewhere and resolved via a bundler, the safelist is activated

**Comment by wongjn** [comment, marked as answer]
Tailwind has no notion of how the scripting in the project works and thus would not know what a component is, let alone whether it is used or not. You could consider updating the `safelist` manually if you do end up using the component.

**Comment by benjaminpreiss** [comment]
I imagine one could write a bundler plugin (e.g. rollup) that does that utilizing [virtual modules](https://vitejs.dev/guide/api-plugin.html#virtual-modules-convention).
The tailwind config used in a project can be extended with:

 ```js
import {safelistComponentsWithBreakpoints} from "virtual:tailwind"
/** @type {import('tailwindcss').Config} */
module.exports = {
  content: [
    './pages/**/*.{html,js}',
    './components/**/*.{html,js}',
  ],
  safelist: [
    ...safelistComponentsWithBreakpoints,
    'bg-red-500',
    'text-3xl',
    'lg:text-4xl',
  ]
  // ...
}
 ```

Then, components could be structured as:

```
global.css
library/
  library.css
  component/
      - `Component.svelte`
      - `component.css`
      - `safelist.js`
  component2/
      - `Component2.svelte`
      - `component.css`
      - `safelist.js`
```

Sample `component.css`:
```css
@layer components {
    .component {
        /*Some styles*/
    }
    .component-medium {
        /*Some styles*/
    }
}
```

`library.css`:
```css
@import "./component/component.css";
@import "./component2/component.css";
```

`global.css` used in the app utilizing tailwind:
```css
@import "tailwindcss/base";

@import "tailwindcss/components";
@import "./library/library.css";

@import "tailwindcss/utilities";
```

And then we can use the safelist based on the `safelist.js` file that is provided in each component:

```js
export default safelist = [/^(?:\w+:)?component-\w+$/] // matches `component` or `md:component` or `sm:component-medium` etc.
```

We then go through all component paths imported from our library within rollup and substitute the `virtual:tailwind` import in the tailwind with a glob import pattern.

In a second processing step the glob import pattern is replaced with imports to only the safelists exported from the components that are imported in the project.

here is a [REPL](https://www.sveltelab.dev/5t6w9cqhpbvj9mg) that covers most of the code, except the rollup plugin
